keps/sig-network/20180612-ipv4-ipv6-dual-stack.md

```diff
@@ -333,36 +337,302 @@ or later. The dual-stack feature requires no changes to this API.
+5. Set `spec.ClusterIPs: ['some ip']` api-server will default `spec.preferDualStack:false` and `spec.ipFamilies: ['ip family of spec.ClusterIP[0]']`. api-server will fail the request if the ip family is not configured on the cluster or the ip is out of cidr range or ip is already allocated to a different service.
+
+The below are examples that follows the above rules.
 
 ```yaml
+# single stack service in cluster that is  configured with --service-cluster-ip-range=<ipv4 cidr> or --service-cluster-ip-range=<ipv4 cidr>,<ipv6 cidr>
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: false # defaulted
+  ipFamilies: # defaulted
+    - IPv4
+  clusterIP: 1.2.3.4 # allocated
+  clusterIPs: # allocated
+    - 1.2.3.4 
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# single stack service in cluster that is configured with --service-cluster-ip-range=<ipv6 cidr> or --service-cluster-ip-range=<ipv6 cidr>,<ipv4 cidr>
 apiVersion: v1
 kind: Service
 metadata:
   name: my-service
 spec:
-  ipFamily: IPv6
+  type: ClusterIP
+  preferDualStack: false # defaulted
+  ipFamilies: # defaulted
+    - IPv6
+  clusterIP: 2001::1 # allocated
+  clusterIPs: # allocated
+    - 2001::1
+  selector:
+    app: MyApp
+  ports:
+    - protocol: TCP
+      port: 80
+      targetPort: 9376
+
+# single stack service with required ipv6, request would fail if cluster is not configured with ipv6 cidr (either first or second cidr)
+apiVersion: v1
+kind: Service
+metadata:
+  name: my-service
+spec:
+  type: ClusterIP
+  preferDualStack: false # defaulted
+  ipFamilies: # set by user
+    - IPv6
+  clusterIP: 2001::1 # allocated
+  clusterIPs:
+    - 2001::1
   selector:
     app: MyApp
   ports:
     - protocol: TCP
       port: 80
       targetPort: 9376
 ```
+> It is implicitly assumed that a service with `spec.preferDualStack: nil` is always a single stack. Users can change this behavior using web hooks.
+
+##### Creating a New Dual Stack Service
+
+Users can create dual stack services according to the following rule
+- If the user require dual stack service (service creation will fail if cluster is not configured for dual stack) then they can do one of the following
+  1. Set `spec.preferDualStack: true` and set `spec.ipFamilies` to any two dual stack ip families in any order.
+  2. set `spec.ClusterIPs: true` and set `spec.ClusterIPs` to any two dual stack ips in any order.
+  3. set `spec.ipFamilies` to any two dual stack ip families in any order.
+  4. Set `spec.CluterIPs` to any two dual stack ip families in any order.
+- If the user prefer dual stack if available (service creation will not fail if the cluster is not configured for dual stack) then they can do one of the following
```

Comment:
It is weird that this is under "Creating a New Dual Stack Service", since it may not result in a dual-stack service...

I think you're torn between half wanting to describe how the code behaves and half wanting to provide end-user use cases, but it gets muddled.

I'm not sure we really need to fix this _now_, but maybe a better organization would be to first give the encyclopedic description of how the apiserver responds to each set of fields, and then after give examples of things like "what happens with a legacy service that doesn't specify any of the new fields", "how to create a service that will get allocated a single IP in a single-stack cluster and dual IPs in a dual-stack cluster", "how to create an IPv4-only service", "how to create an IPv6-primary dual-stack service even in an IPv4-primary dual-stack cluster", "how to manually specify dual cluster IPs", etc. And then you can show both the input yaml and the result after defaulting/allocation.

But that doesn't have to block this merging.

Reply:
I want to do a full edit pass just because it has been patched enough times now :)